### DISCUSSION TITLE
About NFacesList vs. FacetList

### DISCUSSION THREAD

**Discussion by SupposeNot** [SupposeNot/RAMP#180] [Ideas]
Currently FacetList and VertexList are attributes of a maniplex, while NFacesList (because it needs two arguments), is not. I could either:

- Make VertexList and FacetList be operations instead of attributes, or
- Make NFacesList by a FacesListByRanks, which would be a nested list of ALL the faces, sorted into sublists by rank.
What do you think?

**Comment by CunningGabe** [comment]
We have a similar situation for things like NumberOfVertices. At the moment, NumberOfVertices is an attribute, and NumberOfIFaces is an operation. I think that kind of situation is okay - the main difference between an attribute and an operation in this case is just whether the value is saved or not, and it's somewhat reasonable to assume that we more often care about vertices and facets than the other i-faces.

So I guess that's a vote for either keep it the way it is. I guess you could make FacesListByRanks (as you suggested) be an attribute, and then have FacetList, VertexList, and NFacesList be operations that use that attribute - the downside is that then you have to calculate the entire FacesListByRanks even if you only care about a single rank.

There's one more solution I can think of, which is how I handled NumberOfIFaces - though it's a bit messy. There is an FVector attribute, but also there is an internal field fvec which stores all of the entries of the fvector we've calculated so far. So if you ask for NumberOfIFaces, it checks to see if the FVector attribute is already saved, and if so it uses that. If not, it checks the internal fvec to see if that particular NumberOfIFaces is known. If not, then it calculates it and stores it in the internal fvec, and then if the fvec is full, it is propagated out to the FVector attribute.

**Reply by SupposeNot** [reply]
I think the fvec solution is worth keeping in the back pocket, but this is probably a low enough frequency use case that it wouldn't be worth the effort.
